snakemake/shell.py

```diff
@@ -79,6 +80,25 @@ def cleanup(cls):
         with cls._lock:
             cls._processes.clear()
 
+    @classmethod
+    def __get_output_streams(cls, log: list, collect_output):
+        def find_log(prop, name):
+            match = [f for f in log if isinstance(f, _IOFile) and prop.fget(f)]
+            if len(match) == 0:
+                return None
+            if len(match) > 1:
+                raise SyntaxError(name + " allowed on just one log per rule.")
+            makedirs(os.path.dirname(os.path.abspath(match[0])))
+            return open(match[0], "w")
+
+        stdout = find_log(_IOFile.is_stdout, "stdout")
+        stderr = find_log(_IOFile.is_stderr, "stderr")
```

Comment:
It's kind of just right here where I feel like I'm losing separation of concerns, with the `io` objects bleeding into your `shell` class.  It might be justifiable, given that the `shell` class already has a somewhat hazy interface boundary through its use of `inspect.currentframe().f_back.f_locals`.